utils/managed_bytes.hh

```diff
@@ -30,215 +30,213 @@
     }
-    void free_chain(blob_storage* p) noexcept {
+    char* internal_size_slot() {
+        return &_storage[sizeof(_storage) - 1];
+    }
+    char* external_size_slot() {
+        return &_storage[sizeof(_storage) - sizeof(blob_storage::size_type)];
+    }
+    blob_storage::ref_type& first_fragment() {
+        return *reinterpret_cast<blob_storage::ref_type*>(_storage);
+    }
+    const blob_storage::ref_type& first_fragment() const {
+        return *reinterpret_cast<const blob_storage::ref_type*>(_storage);
+    }
+    size_t max_alloc(allocation_strategy& alctr) {
+        return alctr.preferred_max_contiguous_allocation();
+    }
+    void free_chain(blob_storage::ref_type& p) noexcept {
         auto& alctr = current_allocator();
-        while (p) {
-            auto n = p->next;
-            alctr.destroy(p);
+        while (p.has_next()) {
+            blob_storage::ref_type n = p.next();
+            alctr.destroy(p.ptr);
             p = n;
+            // `ptr` can be null iff an allocation failed during construction.
+            if (p.ptr) {
+                p.ptr->backref = &p;
+            }
+        }
+        // `ptr` can be null iff an allocation failed during construction.
+        if (p.ptr) {
+            alctr.destroy(p.ptr);
         }
     }
     bytes_view::value_type& value_at_index(blob_storage::size_type index) {
         if (!external()) {
-            return _u.small.data[index];
+            return reinterpret_cast<bytes_view::value_type&>(_storage[index]);
         }
-        blob_storage* a = _u.ptr;
-        while (index >= a->frag_size) {
-            index -= a->frag_size;
-            a = a->next;
+        const blob_storage::ref_type* a = &first_fragment();
+        while (index >= a->frag_size()) {
+            index -= a->frag_size();
+            a = &a->next();
         }
-        return a->data[index];
+        return a->data()[index];
     }
     std::unique_ptr<bytes_view::value_type[]> do_linearize_pure() const;
 
 public:
-    using size_type = blob_storage::size_type;
     struct initialized_later {};
 
     managed_bytes() {
-        _u.small.size = 0;
+        *internal_size_slot() = 0;
     }
 
-    managed_bytes(const blob_storage::char_type* ptr, size_type size)
+    managed_bytes(const blob_storage::char_type* ptr, blob_storage::size_type size)
         : managed_bytes(bytes_view(ptr, size)) {}
 
     explicit managed_bytes(const bytes& b) : managed_bytes(static_cast<bytes_view>(b)) {}
 
     template <FragmentedView View>
     explicit managed_bytes(View v);
 
-    managed_bytes(initialized_later, size_type size) {
+    managed_bytes(initialized_later, blob_storage::size_type size) {
         memory::on_alloc_point();
         if (size <= max_inline_size) {
-            _u.small.size = size;
+            *internal_size_slot() = size << 1;
         } else {
-            _u.small.size = -1;
+            if (size >= size_t(2) * 1024 * 1024 * 1024) {
```

Comment:
I think this calculation should be based on the number of available bits for storing the size in an automatic way.

Reply:
Right now the number of bits is hardcoded anyway. Changing it to something else than 31 would require some modifications.